internal_transfers/actions/src/simulate/tenderly.ts

```diff
@@ -0,0 +1,109 @@
+import axios from "axios";
+import {
+  SimulationData,
+  SimulationParams,
+  TransactionSimulator,
+} from "./interface";
+import { EventLog } from "../models";
+
+export class TenderlySimulator implements TransactionSimulator {
+  BASE_URL = "https://api.tenderly.co/api";
+  user: string;
+  project: string;
+  accessKey: string;
+
+  constructor(user: string, project: string, accessKey: string) {
+    this.user = user;
+    this.project = project;
+    this.accessKey = accessKey;
+  }
+  async simulate(params: SimulationParams): Promise<SimulationData> {
+    const SIMULATE_URL = `${this.BASE_URL}/v1/account/${this.user}/project/${this.project}/simulate`;
+    const {
+      sender,
+      contractAddress,
+      callData,
+      blockNumber,
+      value = "0",
+    } = params;
+    const response = await axios.post(
+      SIMULATE_URL,
+      {
+        network_id: "1",
+        from: sender,
+        to: contractAddress,
+        input: callData,
+        block_number: blockNumber,
+        gas: 5000000,
+        gas_price: "0",
+        value,
+        // simulation config (tenderly specific)
+        // We have decided to always save simulations...
+        // then we can store simulation IDs and refer back to them
+        save_if_fails: true,
+        save: true,
+        simulation_type: "quick",
+      },
+      {
+        headers: {
+          "X-Access-Key": this.accessKey,
+          "Content-Type": "application/json",
+          Accept: "application/json",
+        },
+      }
+    );
+    if (!isTenderlySimulationResponse(response.data)) {
+      throw Error(`Invalid Response ${JSON.stringify(response.data)}`);
+    }
+    return parseTenderlySimulation(response.data);
+  }
+}
+
+interface TenderlySimulationResponse {
+  transaction: TenderlyTransaction;
+}
+
+interface TenderlyTransaction {
+  transaction_info: TenderlyTransactionInfo;
+  hash: string;
+}
+
+interface TenderlyTransactionInfo {
+  block_number: number;
+  logs: TenderlyTransactionLog[];
+}
+
+interface TenderlyTransactionLog {
+  raw: EventLog;
+}
+
+export function isTenderlySimulationResponse(
+  value: any
+): value is TenderlySimulationResponse {
```

Comment:
Out of curiosity: Why isn't this a boolean?

Reply:
Very good question here! This is some typescript magic I was shown by @anxolin and @tukantje. 

This return type is technically a Boolean, but also performs the type cast!

Reply:
Some context, this is called "type guards" https://www.typescriptlang.org/docs/handbook/advanced-types.html